File: atcoder/trainings/abc319_c.py
import itertools

g = []
for _ in range(3):
    g.append(list(map(int, input().split())))

h = {
    0: [0, 0],
    1: [0, 1],
    2: [0, 2],
    3: [1, 0],
    4: [1, 1],
    5: [1, 2],
    6: [2, 0],
    7: [2, 1],
    8: [2, 2]
}

all_permutations = list(itertools.permutations(range(9)))
all_permutations_cnt = len(all_permutations)
cnt = 0

for pm in all_permutations:
    y1 = []
    y2 = []
    y3 = []
    x1 = []
    x2 = []
    x3 = []
    xy1 = []
    xy2 = []

    for i in range(9):
        a = h[pm[i]]
        if a == [0, 0]:
            y1.append(g[0][0])
            x1.append(g[0][0])
            xy1.append(g[0][0])
        elif a == [0, 1]:
            y2.append(g[0][1])
            x1.append(g[0][1])
        elif a == [0, 2]:
            y3.append(g[0][2])
            x1.append(g[0][2])
            xy2.append(g[0][2])
        elif a == [1, 0]:
            y1.append(g[1][0])
            x2.append(g[1][0])
        elif a == [1, 1]:
            y2.append(g[1][1])
            x2.append(g[1][1])
            xy1.append(g[1][1])
            xy2.append(g[1][1])
        elif a == [1, 2]:
            y3.append(g[1][2])
            x2.append(g[1][2])
        elif a == [2, 0]:
            y1.append(g[2][0])
            x3.append(g[2][0])
            xy2.append(g[2][0])
        elif a == [2, 1]:
            y2.append(g[2][1])
            x3.append(g[2][1])
        elif a == [2, 2]:
            y3.append(g[2][2])
            x3.append(g[2][2])
            xy1.append(g[2][2])

    if ((y1[0] == y1[1] and y1[0] != y1[2]) or
       (y2[0] == y2[1] and y2[0] != y2[2]) or
       (y3[0] == y3[1] and y3[0] != y3[2]) or
       (x1[0] == x1[1] and x1[0] != x1[2]) or
       (x2[0] == x2[1] and x2[0] != x2[2]) or
       (x3[0] == x3[1] and x3[0] != x3[2]) or
       (xy1[0] == xy1[1] and  xy1[0] != xy1[2]) or
       (xy2[0] == xy2[1] and  xy2[0] != xy2[2])):
      cnt += 1

print((all_permutations_cnt - cnt) / float(all_permutations_cnt))

File: atcoder/trainings/abc319_c.rb
# https://atcoder.jp/contests/abc319/tasks/abc319_c
#
# ■考えたこと
#
# ■解説
#
g = Array.new(3)
g[0] = gets.split.map(&:to_i)
g[1] = gets.split.map(&:to_i)
g[2] = gets.split.map(&:to_i)

h = {
  0 => [0, 0],
  1 => [0, 1],
  2 => [0, 2],
  3 => [1, 0],
  4 => [1, 1],
  5 => [1, 2],
  6 => [2, 0],
  7 => [2, 1],
  8 => [2, 2]
}
all = 9 * 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1
cnt = 0
perm = (0..8).to_a.permutation(9).to_a
perm.each do |pm|
  y1 = []
  y2 = []
  y3 = []
  x1 = []
  x2 = []
  x3 = []
  xy1 = []
  xy2 = []
  9.times do |i|
    a = h[pm[i]]
    case a
    when [0, 0]
      y1 << g[0][0]
      x1 << g[0][0]
      xy1 << g[0][0]
    when [0, 1]
      y2 << g[0][1]
      x1 << g[0][1]
    when [0, 2]
      y3 << g[0][2]
      x1 << g[0][2]
      xy2 << g[0][2]
    when [1, 0]
      y1 << g[1][0]
      x2 << g[1][0]
    when [1, 1]
      y2 << g[1][1]
      x2 << g[1][1]
      xy1 << g[1][1]
      xy2 << g[1][1]
    when [1, 2]
      y3 << g[1][2]
      x2 << g[1][2]
    when [2, 0]
      y1 << g[2][0]
      x3 << g[2][0]
      xy2 << g[2][0]
    when [2, 1]
      y2 << g[2][1]
      x3 << g[2][1]
    when [2, 2]
      y3 << g[2][2]
      x3 << g[2][2]
      xy1 << g[2][2]
    end

    next unless (y1[0] == y1[1] && y1[1].nil?.! && y1[1] != y1[2]) ||
                (y2[0] == y2[1] && y2[1].nil?.! && y2[1] != y2[2]) ||
                (y3[0] == y3[1] && y3[1].nil?.! && y3[1] != y3[2]) ||
                (x1[0] == x1[1] && x1[1].nil?.! && x1[1] != x1[2]) ||
                (x2[0] == x2[1] && x2[1].nil?.! && x2[1] != x2[2]) ||
                (x3[0] == x3[1] && x3[1].nil?.! && x3[1] != x3[2]) ||
                (xy1[0] == xy1[1] && xy1[1].nil?.! && xy1[1] != xy1[2]) ||
                (xy2[0] == xy2[1] && xy2[1].nil?.! && xy2[1] != xy2[2])

    cnt += 1
    break
  end
end

puts((all - cnt) / all.to_f)
